### PR TITLE
Bug 5322: Do not leak HttpReply when checking http_reply_access

### PR DESCRIPTION
**Pull request by rousskov**
... as well as response_delay_pool and send_hit directives.

    auto check = clientAclChecklistCreate(...); // sets check->reply
    check->reply = reply; // self-assignment does nothing
    HTTPMSGLOCK(check->reply); // an unwanted/extra lock

When ACLFilledChecklist::reply is already set to X, resetting it to X
should not change HttpReply lock count, but some manual locking code did
not check that "already set" precondition and over-locked reply objects
set to ClientHttpRequest::al::reply by clientAclChecklistFill().

Current known leaks were probably introduced in 2021 commit e227da8 that
started supplying HttpReply to ACLChecklist in clientAclChecklistFill().
The added code locked the reply object correctly, but it was
incompatible with unconditional manual locks in three existing indirect
clientAclChecklistFill() callers (calling clientAclChecklistCreate()).

This change removes all known similar leaks and improves
ACLFilledChecklist API to prevent future similar leaks.

### CODE REVIEW

File: src/HttpHeaderTools.cc
Comment (rousskov) [inline src/HttpHeaderTools.cc:301]: This PR uses new updateAle() in this and similar dangerous `HTTPMSGLOCK(checklist.reply)` places. After these changes, more than a dozen places in Squid code still set ALE::al directly instead of calling updateAle(). They all should eventually be upgraded to updateAle() or similar. To avoid mishaps, ALE::al member itself should made private, affecting dozens of readers. All those large/noisy changes are outside this bug-fixing PR scope.

File: src/http.cc
Comment (rousskov) [inline src/http.cc:775]: In this and similar PR contexts, it is possible that ALE::reply exists and, hence, calling updateReply() is not necessary. However, investigating these hidden dependencies between ALE::reply and job reply fields requires significant effort (especially where a single job deals with multiple reply objects). I did not perform that investigation for this PR, preserving the original "update both ALE and reply" logic (except for trivial cases where the reply object was obviously taken from the ALE object in the same code snippet). I believe that investigation and related further improvements should be viewed as out of scope.

File: src/adaptation/icap/icap_log.cc
Comment (rousskov) [inline src/adaptation/icap/icap_log.cc:64]: In places like this, it is possible that the new updateAle() will supply more information to the checklist than the original reply-setting code did. Hopefully, providing access to additional information is an improvement, but it may affect ACL-driven directives in some corner cases (e.g., a server_cert_fingerprint ACL might now match after getting access to ALE::hier.tcpServer that was not available to that ACL before this updateAle() call).

**Review by rousskov — COMMENTED** [review]
The comments in this review just annotate the diff. They do not request any changes.

**Review by kinkie — COMMENTED** [review]
Thanks for the extensive documentation.
How large is the leak? From the number and location of the removed locks, it seems rather large.

**Comment by rousskov** [comment]
> How large is the leak? From the number and location of the removed locks, it seems rather large.

I do not know the exact answer to your question, but the leak ought to depend on the total response headers size (from hundreds of bytes to tens of kilobytes per transaction?). Squid is leaking HttpReply object (including its HttpHeader member containing header fields and various header parsing caches). In some environments, folks did notice significant (from their point of view) memory growth compared to v5 when evaluating v6, but I do not know specifics.

I speculate that this leak did not attract more attention because relatively few deployments use http_reply_access, send_hit, and response_delay_pool directives.

File: src/acl/FilledChecklist.h
Comment (yadij) [inline src/acl/FilledChecklist.h:69]: This call is implicitly part of reference counting and should clear anything already held when receiving a nil pointer. Otherwise we risk re-using checklists with wrong transaction pointers.

File: src/acl/HttpReqHeader.h
Comment (yadij) [inline src/acl/HttpReqHeader.h:20]: Please do not bundle `HttpHeader` type changes into a change documented as being for `HttpReply`. They are very different objects.

File: src/acl/FilledChecklist.h
Comment (yadij) [inline src/acl/FilledChecklist.h:50]: Please separate these `AccessLogEntry` changes from the `HttpReply` change.

File: src/client_side_reply.cc
Comment (yadij) [reply src/client_side_reply.cc:1839]: Hmm, something seems wrong about that statement. These lines were correctly locking after setting the pointer.

Perhapse you meant that `clientAclChecklistCreate()` already set some value in reply that was not being correctly `HTTPMSGUNLOCK()`ed ?

File: src/acl/FilledChecklist.cc
Comment (yadij) [inline src/acl/FilledChecklist.cc:278]: This breaks refcounting which already handles the nil check properly.
```suggestion
    reply_ = r; // refcounted
```

**Review by yadij — CHANGES_REQUESTED** [review]

**Comment by yadij** [comment]
> Thanks for the extensive documentation. How large is the leak? From the number and location of the removed locks, it seems rather large.

In addition to the thinks @rousskov already mentioned, the `HttpReply` also significantly stores a reference lock on the relevant `HttpRequest` and all its associated headers and caches etc. Which more than doubles the amount of memory used by just the reply message.

So I would expect the leak to be rather significant, if rarely encountered.

File: src/client_side_reply.cc
Comment (rousskov) [reply src/client_side_reply.cc:1839]: > These lines were correctly locking after setting the pointer.

No, they were not. As I tried to explain in the PR description, this code _looks_ OK, but it is buggy:

```C++
foo = bar;
HTTPMSGLOCK(foo); // BUG: Overlocking when foo was bar before the assignment
```

> Perhapse you meant that clientAclChecklistCreate() already set some value in reply that was not being correctly HTTPMSGUNLOCK()ed ?

I did not mean to say that (and did mean to say what I said). However, if we were stuck with manual locks/unlocks, then one could fix this bug by carefully unlocking replyChecklist->reply while resetting it to the same value and locking it again. The end result would be the same -- the final value of reference counter will decrease by one in the problematic case of self-assignment (it will stay the same/unchanged instead of being increased by one as in the current official code):

```C++
// one possible implementation that does not assume anything about prior values:
LOCK(bar); // in case foo is bar
UNLOCK(foo);
foo = bar;
```

Fortunately smart pointers do all this work for us.

**Comment by rousskov** [comment]
> the `HttpReply` also significantly stores a reference lock on the relevant `HttpRequest`

Where? I do not see HttpReply storing an HttpRequest reference;  mgr:mem snapshots I have seen for this leak did not show an increase in HttpRequest objects.

File: src/acl/HttpHeaderData.cc
Comment (rousskov) [reply src/acl/HttpHeaderData.cc]: And those PR 1771 changes are now merged from master into this PR branch (commit 3f2e468c). I am removing S-waiting-for-PR label. This PR continues to wait for @yadij review.

**Comment by rousskov** [comment]
@yadij, please unblock this PR. It has no (known to me) problems and has been waiting for your review for a month.

**Comment by rousskov** [comment]
I have added "Bug 5322" prefix to the PR title. I am not sure this PR fixes _all_ leaks reported (by various people) in that bug report, but it does fix [some](https://bugs.squid-cache.org/show_bug.cgi?id=5322#c19). I do not insist on this prefix; please remove it if you dislike it.

**Review by kinkie — APPROVED** [review]
As it's been successfully run-tested, I'm approving this

**Review by yadij — APPROVED** [review]

**Pull request closed without merging**
